server/jamovi/server/dataset/duckdataset.py

```diff
@@ -172,26 +181,58 @@ def __iter__(self) -> Iterator[DuckColumn]:
         """set a value in the data set"""
-        col = self._columns_by_index[column]
-
-        if isinstance(value, str) and col.data_type is DataType.TEXT and col.has_levels:
+        # Question: Did you want the signature to be changed to take row and column as a tuple?
+        col = self._columns_by_index[column_index]
+        value = self._convert_value_to_raw(value, col)
+        self._write_buffer.stage(CellCoordinate(row_index=row, col_iid=col.iid), value=value)
+
+    def set_values(self, values:dict[CellCoordinate, CellValue]):
+        self._set_values(values, index_name="index")
+
+    def _set_values(self, values:dict[CellCoordinate, CellValue], index_name):
+        
+        # Values written one column at a time base on the assumption that cells 
+        # to be set are not neccessarily adjacent.
+         
+        # Note: Could be improved by updating groups of columns that have 
+        # updates for all the same rows as each other
+        col_iids = set([t[1] for t in values.keys()])
+        for col_i in col_iids:
+            col_updates:dict[int,CellValue] = {index:value for (index, col_j), value in values.items() if col_i==col_j}
+            self._update_column_values(col_i, col_updates, index_name)
+
+    def _update_column_values(self, col, updates:dict[int,CellValue], index_name):
+        col = self._columns_by_iid[col]
+        data_type = col.data_type
+        measure_type = col.measure_type
+        col_iid = col.iid
+        values_line = ",".join([f"({row_index},{self.as_sql_string(value, data_type=data_type, measure_type=measure_type)})" for row_index, value in updates.items()])
+        
+        update_line = f'UPDATE "sheet_data_{self._id}"' 
+        set_line = f'SET "{col_iid}" = t.col_iid'
+        from_line = f'FROM (VALUES {values_line}) AS t(index_col, col_iid)'
+        where_line = f'WHERE "sheet_data_{self._id}".{index_name} = t.index_col'
+        
+        # New newlines included for easier debugging
+        query = f"{update_line}\n{set_line}\n{from_line}\n{where_line}"
+        
+        self._execute(query)
+
+    def commit_set_values(self):
+        self._write_buffer.commit()
+
+    def _convert_value_to_raw(self, value, parent_column: DuckColumn):
+
+        if isinstance(value, str) and parent_column.data_type is DataType.TEXT and parent_column.has_levels:
             try:
-                raw = col.get_value_for_label(value)
+                raw = parent_column.get_value_for_label(value)
             except KeyError:
-                raw = self.column_add_level(col, value)
-            value = raw
-
-        self._execute(
-            f"""
-            UPDATE "sheet_data_{ self._id }"
-            SET "{ col.iid }" = (SELECT $value)
-            WHERE index = $index
-            """,
-            {"value": value, "index": row},
-        )
-        self._cache.clear()
+                raw = self.column_add_level(parent_column, value)
```

Comment:
it's *kinda* frowned upon for a "getter" type function to have side-effects.

i.e. i'm not expecting a function called "convert value to raw" to add new levels into the data set.

Reply:
hmmm, i think this is pre-existing code i moved to it's own function. I took a guess at what it does, but I guess I got it wrong. i'll revisit it

Reply:
it's *kinda* frowned upon, but it's also kinda reasonable ...